pkg/solana/txm/txm.go

```diff
@@ -175,197 +180,197 @@ func (txm *Txm) run() {
-			return solanaGo.Transaction{}, computeUnitErr
-		}
+	// Store tx signature and cancel function
+	if err := txm.txs.New(msg, sig, cancel); err != nil {
+		cancel() // Cancel context when exiting early
+		return solanaGo.Transaction{}, "", solanaGo.Signature{}, fmt.Errorf("failed to save tx signature (%s) to inflight txs: %w", sig, err)
+	}
 
-		// sign tx
-		txMsg, marshalErr := newTx.Message.MarshalBinary()
-		if marshalErr != nil {
-			return solanaGo.Transaction{}, fmt.Errorf("error in soltxm.SendWithRetry.MarshalBinary: %w", marshalErr)
-		}
-		sigBytes, signErr := txm.ks.Sign(ctx, key, txMsg)
-		if signErr != nil {
-			return solanaGo.Transaction{}, fmt.Errorf("error in soltxm.SendWithRetry.Sign: %w", signErr)
-		}
-		var finalSig [64]byte
-		copy(finalSig[:], sigBytes)
-		newTx.Signatures = append(newTx.Signatures, finalSig)
+	txm.lggr.Debugw("tx initial broadcast", "id", msg.id, "fee", msg.cfg.BaseComputeUnitPrice, "signature", sig)
 
-		return newTx, nil
+	// Initialize signature list with initialTx signature. This list will be used to add new signatures and track retry attempts.
+	sigs := &signatureList{}
+	sigs.Allocate()
+	if initSetErr := sigs.Set(0, sig); initSetErr != nil {
+		return solanaGo.Transaction{}, "", solanaGo.Signature{}, fmt.Errorf("failed to save initial signature in signature list: %w", initSetErr)
 	}
 
-	initTx, initBuildErr := buildTx(ctx, baseTx, 0)
-	if initBuildErr != nil {
-		return solanaGo.Transaction{}, "", solanaGo.Signature{}, initBuildErr
-	}
+	// pass in copy of msg (to build new tx with bumped fee) and broadcasted tx == initTx (to retry tx without bumping)
+	txm.done.Add(1)
+	go func() {
+		defer txm.done.Done()
+		txm.retryTx(ctx, msg, initTx, sigs)
+	}()
 
-	// create timeout context
-	ctx, cancel := context.WithTimeout(ctx, msg.cfg.Timeout)
+	// Return signed tx, id, signature for use in simulation
+	return initTx, msg.id, sig, nil
+}
 
-	// send initial tx (do not retry and exit early if fails)
-	sig, initSendErr := txm.sendTx(ctx, &initTx)
-	if initSendErr != nil {
-		cancel()                                                         // cancel context when exiting early
-		txm.txs.OnError(sig, txm.cfg.TxRetentionTimeout(), TxFailReject) //nolint // no need to check error since only incrementing metric here
-		return solanaGo.Transaction{}, "", solanaGo.Signature{}, fmt.Errorf("tx failed initial transmit: %w", initSendErr)
+// buildTx builds and signs the transaction with the appropriate compute unit price.
+func (txm *Txm) buildTx(ctx context.Context, msg pendingTx, retryCount int) (solanaGo.Transaction, error) {
+	// Set compute unit limit if specified
+	if msg.cfg.ComputeUnitLimit != 0 {
+		if err := fees.SetComputeUnitLimit(&msg.tx, fees.ComputeUnitLimit(msg.cfg.ComputeUnitLimit)); err != nil {
+			return solanaGo.Transaction{}, fmt.Errorf("failed to add compute unit limit instruction: %w", err)
+		}
 	}
```

Comment:
I think previously we were adding the limit outside of `buildTx` so think this might introduce a behavior change. Think we would add the limit to `msg.tx` every time we call this method including in the retry method. Not sure if `SetComputeUnitLimit` is idempotent unless your testing confirmed that. Otherwise, think we need to move this back to `sendWithRetry` before we build the tx for the first time. Or we move this below and only add it to the tx copy.